core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -279,13 +297,16 @@ class ReplicaFetcherThread(name: String,
   }
 
   override def buildLeaderEpochRequest(allPartitions: Seq[(TopicPartition, PartitionFetchState)]): Map[TopicPartition, Int] = {
-    val result = allPartitions
+    val partitionEpochOpts = allPartitions
       .filter { case (_, state) => state.isTruncatingLog }
-      .map { case (tp, _) => tp -> epochCache(tp).latestEpoch }.toMap
+      .map { case (tp, _) => tp -> epochCacheOpt(tp) }.toMap
 
-    debug(s"Build leaderEpoch request $result")
+    val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionEpochOpts.partition { case (tp, epochCacheOpt) => epochCacheOpt.nonEmpty }
+    if (partitionsWithoutEpoch.nonEmpty)
+      handlePartitionsWithErrors(partitionsWithoutEpoch.keys)
```

Comment:
Similarly here, it seems it will be cleaner if we return a set of error partitions to the caller and let AbstractFetcherThread call handlePartitionsWithErrors().